AWSAppSyncClient/Internal/AppSyncSubscriptionWithSync.swift

```diff
@@ -0,0 +1,596 @@
+            self?.baseQueryHandler(result, error)
+            dispatchGroup.leave()
+        }
+        dispatchGroup.wait()
+    }
+
+    /// Load base query results from service, forcing a bypass of the local cache
+    private func runBaseQueryFromService() -> Bool {
+        guard let appSyncClient = appSyncClient else {
+            return false
+        }
+
+        var success: Bool = true
+
+        AppSyncLog.info("Refreshing base query results from service")
+
+        let semaphore = DispatchSemaphore(value: 0)
+
+        let networkFetchTime = Date()
+
+        appSyncClient.fetch(query: baseQuery, cachePolicy: .fetchIgnoringCacheData) { [weak self] (result, error) in
+
+            defer {
+                semaphore.signal()
+            }
+
+            guard let self = self else {
+                return
+            }
+
+            // call customer if successful or error
+            // return false to parent if failed
+            if error == nil {
+                success = true
+            } else if result != nil {
+                success = true
+            } else {
+                success = false
+            }
+
+            self.baseQueryHandler(result, error)
+        }
+
+        semaphore.wait()
+
+        if success {
+            AppSyncLog.debug("Base query refresh from service successful. Updating last sync time.")
+            self.setLastSyncTime(date: networkFetchTime)
+        }
+
+        return success
+    }
+
+    // MARK: - Subscription handling
+
+    /// Initializes a subscription listener for `subscription`. This method returns after either receiving
+    /// a connect callback, or an errored result.
+    ///
+    /// - Returns: true if the subscription was successfully started, or if `subscription` is nil.
+    private func startSubscription() -> Bool {
+        var success: Bool? = nil
+        guard let subscription = subscription, let appSyncClient = appSyncClient else {
+            return true
+        }
+
+        AppSyncLog.info("Starting subscription")
+
+        let semaphore = DispatchSemaphore(value: 0)
+
+        var oldSubscriptionWatcher = subscriptionWatcher
+        let connectCallback = {
+            oldSubscriptionWatcher?.cancel()
+            oldSubscriptionWatcher = nil
+
+            // Guard against multiple invocations of the connect callback
+            if success == nil {
+                success = true
+                semaphore.signal()
+            }
+        }
+
+        let resultHandler: SubscriptionResultHandler<Subscription> = { [weak self] (result, transaction, error) in
+            // `subscribeWithConnectCallback` invokes `resultHandler` with an error if it encounters an error
+            // during the connect phase. Handle that here by updating the success flag and signalling the
+            // semaphore to end the method.
+            if error as? AWSAppSyncSubscriptionError != nil {
```

Comment:
I think we should check for any error (not just subscription error)

Reply:
I think it may be good to have 'retryCount' limit. So if there're `errors.count` > `retryCount`. Then restart of subscription won't happen anymore and e.g. should be done manually.

Reply:
> I think we should check for any error (not just subscription error)

Since we have a `success == nil` check, that makes sense.

> I think it may be good to have 'retryCount' limit. So if there're `errors.count` > `retryCount`. Then restart of subscription won't happen anymore and e.g. should be done manually.

I'd rather do that as part of our TODO to handle subscription setup separately from sync. Currently, `startSubscription` is only ever invoked from `performSync`, because we don't have good lifecycle and network state awareness on subscriptions. Once we tease apart subscription handling from sync handling, having separately-configured and managed retry strategies for sync and subscription setup will be appropriate. 

Regarding "should be done manually": Since the overall consistency of the system is pretty tightly interwoven, and depends on quite a few moving parts, what would your view of a recovery mechanism look like?

Reply:
I think that the best solution is giving API user to control retry count and when to retry, e.g. in case of FRP we can have something like
```swift
appSyncClient
   .sync(...)
   .retry(count: 5)
   .flatMapError(...)
   .subscribe()
```

But this's not what AppSync API needs to introduce, it's already Rx. But we're not in FRP / Rx world.

So here's my opinion:
 1. `AppSyncSubscriptionWithSync` should start / stop subscriptions only when reachability events occur.
 2. start / stop count should be a part `SyncConfiguration` with default value, like `let reachabilityRetryCount: Int = .max`.
 2. `AppSyncSubscriptionWithSync` should check reachability and wait (not restart as now) while host is reachable according to the current client settings.
 3. Some error should be forwarded to outer layer when `reachabilityRetryCount` reached.
 4. Other than Reachability errors should be forwarded to the outer layer and user should decide what to do.
 5. AWSAppSync errors should be revised, since it's wrapped strings in general. So that it's difficult and unsafe to match concrete error.

Reply:
I've got some thoughts on the overall pattern, but let's move the retry handling discussion off to another thread, so we don't pollute the intent of this PR.